File: app/src/main/java/party/qwer/iris/ObserverHelper.java
package party.qwer.iris;

import org.json.JSONObject;
import org.json.JSONException;
import org.json.JSONArray;

import java.util.Collections;
import java.util.Map;
import java.io.BufferedReader;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import android.database.Cursor;
import android.database.sqlite.SQLiteException;
import java.net.HttpURLConnection;
import java.io.OutputStream;
import java.net.URL;
import java.util.Objects;
import java.util.HashMap;
import java.text.SimpleDateFormat;
import java.util.TimeZone;
import java.util.Date;
import java.util.LinkedList;

public class ObserverHelper {
    private long lastLogId = 0;
    private final LinkedList<Map<String, Object>> lastDecryptedLogs = new LinkedList<>(); // Store last decrypted logs
    private static final int MAX_LOGS_STORED = 50; // Maximum number of logs to store

    private String makePostData(String decMsg, String room, String sender, JSONObject js) throws JSONException {
        JSONObject data = new JSONObject();
        data.put("msg", decMsg);
        data.put("room", room);
        data.put("sender", sender);
        data.put("json", js);
        return data.toString();
    }

    public void checkChange(KakaoDB db) {
        if (lastLogId == 0) {
            Map<String, Object> lastLog = db.logToDict(0);
            if (lastLog != null && lastLog.containsKey("_id")) {
                lastLogId = Long.parseLong((String) Objects.requireNonNull(lastLog.get("_id")));
            } else {
                lastLogId = 0;
            }
            System.out.println("Initial lastLogId: " + lastLogId);
            return;
        }

        String countSql = "select count(*) from chat_logs where _id > ?";
        Cursor countRes = null;
        int newLogCount = 0;
        try {
            String[] selectionCountArgs = {String.valueOf(lastLogId)};
            countRes = db.getConnection().rawQuery(countSql, selectionCountArgs);
            if (countRes != null && countRes.moveToNext()) {
                newLogCount = countRes.getInt(0);
            }
        } catch (SQLiteException e) {
            System.err.println("SQL error in checkChange (count query): " + e.getMessage());
        } finally {
            if (countRes != null) {
                countRes.close();
            }
        }

        if (newLogCount > 0) {
            System.out.println("Detected " + newLogCount + " new log(s). Processing...");
            String sql = "select * from chat_logs where _id > ? order by _id asc";
            Cursor res = null;
            try {
                String[] selectionArgs = {String.valueOf(lastLogId)};
                res = db.getConnection().rawQuery(sql, selectionArgs);
                List<String> description = new ArrayList<>();
                if (res.getColumnNames() != null) {
                    Collections.addAll(description, res.getColumnNames());
                }


                while (res.moveToNext()) {
                    long currentLogId = res.getLong(res.getColumnIndexOrThrow("_id"));
                    if (currentLogId > lastLogId) {
                        lastLogId = currentLogId;
                        int encType;
                        String origin = "";
                        try {
                            JSONObject v = new JSONObject(res.getString(res.getColumnIndexOrThrow("v")));
                            encType = v.getInt("enc");
                            origin = v.getString("origin");
                        } catch (JSONException e) {
                            System.err.println("Error parsing 'v' JSON for encType: " + e.getMessage());
                            encType = 0;
                        }

                        if (origin.equals("SYNCMSG") || origin.equals("MCHATLOGS")){
                            continue;
                        }


                        JSONObject logJson = new JSONObject();
                        for (String columnName : description) {
                            try {
                                logJson.put(columnName, res.getString(res.getColumnIndexOrThrow(columnName)));
                            } catch (JSONException e) {
                                System.err.println("JSONException while adding log data to JSON object: " + e.getMessage());
                            }
                        }


                        String enc_msg = res.getString(res.getColumnIndexOrThrow("message"));
                        String enc_attachment = res.getString(res.getColumnIndexOrThrow("attachment"));
                        long user_id = res.getLong(res.getColumnIndexOrThrow("user_id"));



                        String dec_msg;
                        String dec_attachment;
                        try {
                            if (enc_msg == null || enc_msg.isEmpty() || enc_msg.equals("{}")){
                                dec_msg = "{}}";
                            } else {
                                dec_msg = KakaoDecrypt.decrypt(encType, enc_msg, user_id);
                                logJson.put("message", dec_msg);
                            }
                            if (enc_attachment == null || enc_attachment.isEmpty() || enc_attachment.equals("{}") || res.getColumnIndexOrThrow("attachment") == -1){
                                dec_attachment = "{}";
                            } else {
                                dec_attachment = KakaoDecrypt.decrypt(encType, enc_attachment, user_id);
                            }
                            logJson.put("attachment", dec_attachment);
                        } catch (Exception e) {
                            System.err.println("Decryption error for logId " + currentLogId + ": " + e);
                            dec_msg = enc_msg;
                        }

                        long chat_id = res.getLong(res.getColumnIndexOrThrow("chat_id"));
                        String[] userInfo = db.getUserInfo(chat_id, user_id);
                        String room = userInfo[0];
                        String sender = userInfo[1];
                        if (room.equals(Configurable.getInstance().getBotName())) {
                            room = sender;
                        }
                        String postData;
                        try {
                            postData = makePostData(dec_msg, room, sender, logJson);
                            sendPostRequest(postData);
                            // Store decrypted log for config page
                            storeDecryptedLog(res, dec_msg); // Call storeDecryptedLog here
                        } catch (JSONException e) {
                            System.err.println("JSON error creating post data: " + e.getMessage());
                        }
                        System.out.println("New message from " + sender + " in " + room + ": " + dec_msg);

                    }
                }
            } catch (SQLiteException e) {
                System.err.println("SQL error in checkChange (data query): " + e.getMessage());
            } finally {
                if (res != null) {
                    res.close();
                }
            }
        }
    }

    private synchronized void storeDecryptedLog(Cursor cursor, String decryptedMessage) {
        Map<String, Object> logEntry = new HashMap<>();
        logEntry.put("_id", cursor.getString(cursor.getColumnIndexOrThrow("_id")));
        logEntry.put("chat_id", cursor.getString(cursor.getColumnIndexOrThrow("chat_id")));
        logEntry.put("user_id", cursor.getString(cursor.getColumnIndexOrThrow("user_id")));
        logEntry.put("message", decryptedMessage);
        long createdAtTimestamp = cursor.getLong(cursor.getColumnIndexOrThrow("created_at"));
        Date date = new Date(createdAtTimestamp * 1000L);
        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
        sdf.setTimeZone(TimeZone.getTimeZone("GMT+9"));
        logEntry.put("created_at", sdf.format(date));

        lastDecryptedLogs.addFirst(logEntry); // Add to the front as it's reverse chronological
        if (lastDecryptedLogs.size() > MAX_LOGS_STORED) {
            lastDecryptedLogs.removeLast(); // Remove the oldest log if over limit
        }
    }


    public List<Map<String, Object>> getLastChatLogs() {
        return new ArrayList<>(lastDecryptedLogs); // Return a copy to avoid external modification
    }


    // Modified sendPostRequest to fetch endpoint each time
    private void sendPostRequest(String jsonData) {
        String urlStr = Configurable.getInstance().getWebServerEndpoint();
        System.out.println("Sending HTTP POST request to: " + urlStr);
        System.out.println("JSON Data being sent: " + jsonData);
        try {
            HttpURLConnection con = getHttpURLConnection(jsonData, urlStr);

            int responseCode = con.getResponseCode();
            System.out.println("HTTP Response Code: " + responseCode);

            try (BufferedReader br = new BufferedReader(
                    new java.io.InputStreamReader(con.getInputStream(), StandardCharsets.UTF_8))) {
                StringBuilder response = new StringBuilder();
                String responseLine;
                while ((responseLine = br.readLine()) != null) {
                    response.append(responseLine.trim());
                }
                String responseBody = response.toString();
                System.out.println("HTTP Response Body: " + responseBody);
            } catch (IOException e) {
                System.err.println("Error reading HTTP response body: " + e.getMessage());
            }
            con.disconnect();

        } catch (IOException e) {
            System.err.println("IO error sending POST request: " + e.getMessage());
        }
    }

    private static HttpURLConnection getHttpURLConnection(String jsonData, String urlStr) throws IOException {
        URL url = new URL(urlStr);
        HttpURLConnection con = (HttpURLConnection) url.openConnection();
        con.setRequestMethod("POST");
        con.setRequestProperty("Content-Type", "application/json");
        con.setRequestProperty("Accept", "application/json");
        con.setDoOutput(true);

        try (OutputStream os = con.getOutputStream()) {
            byte[] input = jsonData.getBytes(StandardCharsets.UTF_8);
            os.write(input, 0, input.length);
        }
        return con;
    }
}

File: app/src/main/java/party/qwer/iris/ObserverHelper.kt
package party.qwer.iris

import android.database.Cursor
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.runBlocking
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.io.IOException
import java.security.MessageDigest
import java.util.LinkedList
import java.util.concurrent.Executors
import kotlin.collections.set

class ObserverHelper(
    private val db: KakaoDB, private val wsBroadcastFlow: MutableSharedFlow<String>
) {
    private val checkpointTracker = LogCheckpointTracker()
    private val lastDecryptedLogs = LinkedList<Map<String, String?>>()
    private val httpRequestExecutor = Executors.newFixedThreadPool(8)
    private val okHttpClient = OkHttpClient()

    fun checkChange(db: KakaoDB) {
        if (checkpointTracker.checkpoint() == 0L) {
            val initialLogId = getLastLogIdFromDB()
            checkpointTracker.reset(initialLogId)
            println("Initial lastLogId: $initialLogId")
            return
        }

        val newLogCount = getNewLogCountFromDB()
        if (newLogCount <= 0) {
            return
        }

        println("Detected $newLogCount new log(s). Processing...")

        db.connection.rawQuery(
            "SELECT * FROM chat_logs WHERE _id > ? ORDER BY _id ASC",
            arrayOf(checkpointTracker.checkpoint().toString())
        ).use { cursor ->
            while (cursor.moveToNext()) {
                processCursorRow(cursor)
            }
        }
    }

    private fun processCursorRow(cursor: Cursor) {
        val columnNames = cursor.columnNames
        val currentLogId = cursor.getLong(columnNames.indexOf("_id"))

        if (checkpointTracker.shouldSkip(currentLogId)) {
            return
        }

        val chatId = cursor.getLongOrNull(columnNames, "chat_id")
        val userId = cursor.getLongOrNull(columnNames, "user_id")
        val messageType = cursor.getStringOrNull(columnNames, "type")
        val message = cursor.getStringOrNull(columnNames, "message")
        val attachment = cursor.getStringOrNull(columnNames, "attachment")
        val supplement = cursor.getStringOrNull(columnNames, "supplement")
        val metadata = RowMetadata(
            logId = currentLogId,
            chatId = chatId,
            userId = userId,
            messageType = messageType,
            messageLength = message?.length ?: 0,
            attachmentLength = attachment?.length ?: 0,
            supplementLength = supplement?.length ?: 0
        )

        try {
            processRow(cursor, columnNames, metadata)
            checkpointTracker.markCompleted(currentLogId)
        } catch (e: Exception) {
            logRowFailure(metadata, "row", e)
        }
    }

    private fun processRow(cursor: Cursor, columnNames: Array<String>, metadata: RowMetadata) {
        val vRaw = cursor.getStringOrNull(columnNames, "v") ?: "{}"
        val v = try {
            JSONObject(vRaw)
        } catch (e: Exception) {
            logRowFailure(metadata, "v_parse", e)
            JSONObject("{}")
        }
        val enc = v.optInt("enc", 0)
        val origin = v.optString("origin", "")

        if (origin == "SYNCMSG" || origin == "MCHATLOGS") {
            checkpointTracker.markCompleted(metadata.logId)
            return
        }

        val chatId = metadata.chatId ?: 0L
        val userId = metadata.userId ?: 0L
        val messageType = metadata.messageType ?: ""

        var message = cursor.getStringOrNull(columnNames, "message") ?: ""
        var attachment = cursor.getStringOrNull(columnNames, "attachment") ?: "{}"
        val threadId = cursor.getStringOrNull(columnNames, "thread_id")

        var supplement = "{}"
        try {
            supplement = cursor.getStringOrNull(columnNames, "supplement") ?: "{}"
            if (supplement.isNotEmpty() && supplement != "{}") {
                supplement = KakaoDecrypt.decrypt(enc, supplement, userId)
            }
        } catch (e: Exception) {
            logRowFailure(metadata, "supplement_decrypt", e)
            supplement = "{}"
        }

        try {
            if (message.isNotEmpty() && message != "{}") {
                message = KakaoDecrypt.decrypt(enc, message, userId)
            }
        } catch (e: Exception) {
            logRowFailure(metadata, "message_decrypt", e)
        }

        try {
            if ((message.contains("선물") && messageType == "71") || attachment == "null") {
                attachment = "{}"
            } else if (attachment.isNotEmpty() && attachment != "{}") {
                attachment = KakaoDecrypt.decrypt(enc, attachment, userId)
            }
        } catch (e: Exception) {
            logRowFailure(metadata, "attachment_decrypt", e)
            attachment = "{}"
        }

        storeDecryptedLog(cursor, message)

        val raw = mutableMapOf<String, String?>()
        val attachmentMap = getStringJsonToMap(attachment, metadata, "attachment_parse").apply {
            this["src_isThread"] = false
        }
        val supplementMap = getStringJsonToMap(supplement, metadata, "supplement_parse")

        for ((idx, columnName) in columnNames.withIndex()) {
            when (columnName) {
                "message" -> raw[columnName] = message
                "attachment" -> raw[columnName] = attachment
                "supplement" -> raw[columnName] = supplement
                else -> raw[columnName] = cursor.getString(idx)
            }
        }

        if (
            threadId.isNullOrEmpty() &&
            supplementMap.getOrDefault("threadId", "") != "" &&
            attachmentMap.getOrDefault("src_logId", "") == "" &&
            messageType == "1"
        ) {
            attachmentMap["src_logId"] = supplementMap.getOrDefault("threadId", "")
            attachmentMap["src_isThread"] = true
        } else if (!threadId.isNullOrEmpty() && messageType == "1") {
            attachmentMap["src_logId"] = threadId.toLongOrNull() ?: threadId
            attachmentMap["src_isThread"] = true
        }

        raw["attachment"] = JSONObject(attachmentMap as Map<*, *>).toString()
        raw["supplement"] = JSONObject(supplementMap as Map<*, *>).toString()

        val (roomName, senderName) = resolveChatInfo(chatId, userId, metadata)
        val data = JSONObject(
            mapOf(
                "msg" to message,
                "room" to roomName,
                "sender" to senderName,
                "json" to raw
            )
        ).toString()

        runBlocking {
            wsBroadcastFlow.emit(data)
        }

        if (Configurable.webServerEndpoint.isNotEmpty()) {
            httpRequestExecutor.execute {
                sendPostRequest(data)
            }
        }
    }

    private fun resolveChatInfo(chatId: Long, userId: Long, metadata: RowMetadata): Pair<String, String> {
        var roomName: String? = null
        var senderName: String? = null

        try {
            val chatInfo = db.getChatInfo(chatId, userId)
            roomName = chatInfo.getOrNull(0)
            senderName = chatInfo.getOrNull(1)
        } catch (e: Exception) {
            logRowFailure(metadata, "chat_info_lookup", e)
        }

        if (senderName.isNullOrEmpty()) {
            try {
                val rawKey = "person_${chatId}:${userId}"
                val md = MessageDigest.getInstance("SHA-256")
                val hashedId = md.digest(rawKey.toByteArray()).joinToString("") { "%02x".format(it) }

                val fallbackInfo = NamesDB.getName(hashedId)
                if (fallbackInfo != null) {
                    senderName = fallbackInfo.first
                    if (roomName.isNullOrEmpty()) {
                        roomName = fallbackInfo.second
                    }
                }
            } catch (e: Exception) {
                logRowFailure(metadata, "namesdb_lookup", e)
            }
        }

        val fallbackSender = senderName?.takeIf { it.isNotBlank() } ?: "<unknown:$userId>"
        val fallbackRoom = roomName?.takeIf { it.isNotBlank() }
            ?: senderName?.takeIf { it.isNotBlank() }
            ?: "chat:$chatId"

        return fallbackRoom to fallbackSender
    }

    private fun getLastLogIdFromDB(): Long {
        val lastLog = db.logToDict(0)
        return lastLog["_id"]?.toLongOrNull() ?: 0
    }

    private fun getStringJsonToMap(
        data: String?,
        metadata: RowMetadata,
        stage: String
    ): MutableMap<String, Any?> {
        if (data.isNullOrBlank() || data == "{}") {
            return HashMap()
        }

        return try {
            val object_ = JSONObject(data)
            val map: MutableMap<String, Any?> = HashMap()

            val keys: MutableIterator<String> = object_.keys()
            while (keys.hasNext()) {
                val key = keys.next()
                val value: Any? = object_.get(key)
                map[key] = value
            }

            map
        } catch (e: Exception) {
            logRowFailure(metadata, stage, e)
            HashMap()
        }
    }

    private fun getNewLogCountFromDB(): Int {
        val res = db.executeQuery(
            "select count(*) as cnt from chat_logs where _id > ?",
            arrayOf(checkpointTracker.checkpoint().toString())
        )
        return res[0]["cnt"]?.toIntOrNull() ?: 0
    }

    @Synchronized
    private fun storeDecryptedLog(cursor: Cursor, decryptedMessage: String?) {
        val logEntry: MutableMap<String, String?> = HashMap()
        logEntry["_id"] = cursor.getString(cursor.getColumnIndexOrThrow("_id"))
        logEntry["chat_id"] = cursor.getString(cursor.getColumnIndexOrThrow("chat_id"))
        logEntry["user_id"] = cursor.getString(cursor.getColumnIndexOrThrow("user_id"))
        logEntry["message"] = decryptedMessage
        logEntry["created_at"] = cursor.getString(cursor.getColumnIndexOrThrow("created_at"))

        lastDecryptedLogs.addFirst(logEntry)
        if (lastDecryptedLogs.size > MAX_LOGS_STORED) {
            lastDecryptedLogs.removeLast()
        }
    }

    @Synchronized
    fun getRecentDecryptedLogs(limit: Int): List<Map<String, String?>> {
        return lastDecryptedLogs.take(limit)
    }

    val lastChatLogs: List<Map<String, String?>>
        @Synchronized get() = lastDecryptedLogs

    private fun sendPostRequest(jsonData: String) {
        try {
            val requestBody = jsonData.toRequestBody("application/json; charset=utf-8".toMediaType())
            val request = Request.Builder()
                .url(Configurable.webServerEndpoint)
                .post(requestBody)
                .build()

            okHttpClient.newCall(request).execute().use { response ->
                if (!response.isSuccessful) {
                    println("Failed to send POST request. Response code: ${response.code}")
                }
            }
        } catch (e: IOException) {
            println("Error sending POST request: ${e.message}")
        }
    }

    private fun logRowFailure(metadata: RowMetadata, stage: String, error: Exception) {
        System.err.println(
            "Row processing failed" +
                " _id=${metadata.logId}" +
                " chat_id=${metadata.chatId ?: "null"}" +
                " user_id=${metadata.userId ?: "null"}" +
                " type=${metadata.messageType ?: "null"}" +
                " message_len=${metadata.messageLength}" +
                " attachment_len=${metadata.attachmentLength}" +
                " supplement_len=${metadata.supplementLength}" +
                " stage=$stage" +
                " error=${error.javaClass.simpleName}:${error.message}"
        )
    }

    private data class RowMetadata(
        val logId: Long,
        val chatId: Long?,
        val userId: Long?,
        val messageType: String?,
        val messageLength: Int,
        val attachmentLength: Int,
        val supplementLength: Int
    )

    companion object {
        private const val MAX_LOGS_STORED = 50
    }
}

private fun Cursor.getStringOrNull(columnNames: Array<String>, columnName: String): String? {
    val index = columnNames.indexOf(columnName)
    if (index == -1 || isNull(index)) {
        return null
    }
    return getString(index)
}

private fun Cursor.getLongOrNull(columnNames: Array<String>, columnName: String): Long? {
    val index = columnNames.indexOf(columnName)
    if (index == -1 || isNull(index)) {
        return null
    }
    return getLong(index)
}
